From d82c94d8e27ce72eb45dfb23ccc9271493dad98d Mon Sep 17 00:00:00 2001
From: Brian Behlendorf <behlendorf1@llnl.gov>
Date: Fri, 30 Oct 2015 08:47:09 -0700
Subject: [PATCH 1/3] Handle AttributeError exception in run_cleanups()

When for whatever reason a slave isn't created then slavebuilder.slave
will be None for the cleanup callbacks.  Cleanly handle this case when
running the callbacks.  This was observed in a configuration when used
ec2 latent buildslaves.

2015-10-28 10:03:53-0700 [-] while preparing slavebuilder:
        Traceback (most recent call last):
        Failure: twisted.internet.defer.TimeoutError:
...
cleanups.append(lambda: slavebuilder.slave.releaseLocks())
exceptions.AttributeError: 'NoneType' object has no attribute 'releaseLocks'

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
---
 master/buildbot/process/builder.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/master/buildbot/process/builder.py b/master/buildbot/process/builder.py
index 85d7813..5aa265e 100644
--- a/master/buildbot/process/builder.py
+++ b/master/buildbot/process/builder.py
@@ -316,6 +316,8 @@ class Builder(util_service.ReconfigurableServiceMixin,
                 while cleanups:
                     fn = cleanups.pop()
                     fn()
+            except AttributeError:
+                log.err("failed to start slave %s" % (slavebuilder))
             except Exception:
                 log.err(failure.Failure(), "while running %r" % (run_cleanups,))
 
-- 
1.9.1

